### DISCUSSION TITLE
Integration of a hardware accelerator for CKKS (fully homomorphic encryption) into the NEORV32 CFS.

### DISCUSSION THREAD

**Discussion by Sam-Vervaeck** [stnolting/neorv32#1073] [Show and tell]
Hey @stnolting! It has been a while since I posted a new discussion topic but here I am again. This may well be my final one since I am finished with my master's thesis project.

I recently uploaded my project code to GitHub to make it open-source available: [NEORV32_AlohaHE_thesis](https://github.com/Sam-Vervaeck/NEORV32_AlohaHE_thesis). You can find a short description of how I used the NEORV32 in that repo. The CFS came in very handy!

I want to mention again how grateful I am for your support of the NEORV32, for being patient with new users and for answering all my questions. You also got a shout-out in the introduction of my thesis text :).

For the lazy among us I include the same short description here:

> During my Master's thesis of 2023-2024 at the COSIC group at the KU Leuven, I had a topic on the CKKS algorithm for fully homomorphic encryption (FHE). I made a demo to show the integration of a hardware accelerator for CKKS and a RISC-V core. For the accelerator I chose the state-of-the-art [Aloha-HE](https://github.com/flokrieger/Aloha-HE) accelerator from Florian Krieger et al. at that time. For the RISC-V core I chose the [NEORV32](https://github.com/stnolting/neorv32/tree/main) by Stephan Nolting.
> 
> This repository contains a Vivado project showcasing a CKKS encryption demo. The Aloha-HE accelerator is integrated into the Custom Fuction Subsystem (CFS) of the NEORV32 and has an input and output channel that propagates all the way to the toplevel op the core. A block diagram contains the NEORV32 and a BRAM which stores the test data for the CKKS computations through a .coe file initialisation. The program execution is controlled by NEORV32 and the CKKS computations are performed by Aloha-HE by offloading the data to the CFS block containing the accelerator.

**Comment by stnolting** [comment]
Hey @Sam-Vervaeck.

Coincidentally, I'm also in the process of learning more about homomorphic encryption - so thank you very much for sharing this! :+1:

Will your actual thesis be published as well? I would be really interested in reading it.

**Reply by Sam-Vervaeck** [reply]
I'll ask about the confidentiality.

**Reply by stnolting** [reply]
Cool, thanks!

**Comment by Sam-Vervaeck** [comment]
Hey @stnolting ,

I am happy to say I can share it with you, apparently I have full control over my text :). I will attach the PDF to this message.

A bit of context as well:

- I took out some unrelated KUL pages from the text.
- Sadly, I had to submit my text already at the end of August while I was still working on my final design. I got that better design working just in time for my thesis defense presentation in September but it is not included in this text. I added a snippet of my defense presentation that explains the final design and the final performance and area results.
- The best (most useful) parts of my text are the introduction into FHE and the CKKS algorithm if you are learning about those.

[thesis_github.pdf](https://github.com/user-attachments/files/17636943/thesis_github.pdf)
[final_design.pdf](https://github.com/user-attachments/files/17637055/final_design.pdf)

**Reply by stnolting** [reply]
Hey @Sam-Vervaeck!

For some reason I completely overlooked this post... Sorry for that. 🙈

Anyway, thank you very much for sharing! I just scrolled through your thesis and it looks really great! If you are okay with it, I would like to share it with a colleague (we are investigating on isomorphic encryption..).

Oh, and congratulations for finishing your masters degree! 🎉 

\cc @crolfes

**Reply by Sam-Vervaeck** [reply]
Hi!

No worries, thank you for the compliment :). Feel free to share the text! It is nice to know somebody might learn something from what I wrote.
